tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php

```diff
@@ -0,0 +1,640 @@
+            throw new RuntimeException(sprintf(
+                'Cart rule reduction tax flag "%s" is not expected',
+                var_export($isTaxIncluded)
+            ));
+        }
+
+        $this->assertDiscountApplicationTypeIsValid($cartRule, $discountApplicationType);
+    }
+
+    /**
+     * @Given /^(it) should give a percentage reduction of "([^"]+)" which (excludes|includes) discounted products and (applies to (order without shipping|specific product|cheapest product|selected products))$/
+     */
+    public function itShouldGivePercentageReduction(
+        CartRule $cartRule,
+        $percentage,
+        bool $includesDiscountedProducts,
+        string $discountApplicationType
+    ) {
+        if (!$this->areNumbersEqual($cartRule->reduction_percent, $percentage)) {
+            throw new RuntimeException(sprintf(
+                'Cart rule reduction percentage "%s" is not expected',
+                $percentage
+            ));
+        }
+
+        if ((bool) $cartRule->reduction_exclude_special !== !$includesDiscountedProducts) {
+            throw new RuntimeException(sprintf(
+                'Cart rule reduction_exclude_special flag "%s" is not expected',
+                var_export($includesDiscountedProducts)
+            ));
+        }
+
+        $this->assertDiscountApplicationTypeIsValid($cartRule, $discountApplicationType);
+    }
+
+    /**
+     * @Transform /^(active from|active until|quantity per user|partial use|status|highlight in cart)$/
+     */
+    public function getMappedProperty(string $property): string
+    {
+        $map = [
+            'active from' => 'date_from',
+            'active until' => 'date_to',
+            'quantity per user' => 'quantity_per_user',
+            'partial use' => 'partial_use',
+            'status' => 'active',
+            'highlight in cart' => 'highlight',
+        ];
+
+        return $map[$property] ?? $property;
+    }
+
+    /**
+     * @Transform /^applies to ([^"]+)$/
+     */
+    public function getMappedDiscountApplicationType(string $type)
+    {
+        $map = [
+            'order without shipping' => DiscountApplicationType::ORDER_WITHOUT_SHIPPING,
+            'specific product' => DiscountApplicationType::SPECIFIC_PRODUCT,
+            'cheapest product' => DiscountApplicationType::CHEAPEST_PRODUCT,
+            'selected products' => DiscountApplicationType::SELECTED_PRODUCTS,
+        ];
+
+        return $map[$type] ?? $type;
+    }
+
+    /**
+     * Asserts that discount application type is valid for given cart rule
+     *
+     * @param CartRule $cartRule
+     * @param string $discountApplicationType
+     */
+    private function assertDiscountApplicationTypeIsValid(CartRule $cartRule, string $discountApplicationType)
+    {
+        switch ($discountApplicationType) {
+            case DiscountApplicationType::ORDER_WITHOUT_SHIPPING:
+                $hasError = 0 !== (int) $cartRule->reduction_product;
```

Comment:
I see these hardcoded values for the seconds time, would it make sense to add  constants?

Reply:
maybe, somewhere in the `Adapter` namespace as it's coming from legacy